packages/firestore/scripts/assertion-id-tool.ts

```diff
@@ -0,0 +1,372 @@
+  const foundCalls: CallSiteInfo[] = [];
+
+  for (const filePath of filePaths) {
+    try {
+      // Read the file content
+      const sourceText = fs.readFileSync(filePath, "utf8");
+
+      // Create the SourceFile AST node
+      const sourceFile = ts.createSourceFile(
+        path.basename(filePath), // Use basename for AST node name
+        sourceText,
+        ts.ScriptTarget.ESNext, // Or your project's target
+        true, // Set parent pointers
+        ts.ScriptKind.Unknown // Detect TS vs TSX automatically
+      );
+
+      // Define the visitor function
+      const visit = (node: ts.Node) :void => {
+        // Check if the node is a CallExpression (e.g., myFunction(...))
+        if (ts.isCallExpression(node)) {
+          let functionName: string | null = null;
+          const expression = node.expression;
+
+          // Check if the call is directly to an identifier (e.g., fail())
+          if (ts.isIdentifier(expression)) {
+            functionName = expression.text;
+          }
+
+          // If we found a function name, and it's one we're looking for
+          if (functionName && targetFunctionNames.has(functionName)) {
+            // Get line and character number
+            const { line, character } = ts.getLineAndCharacterOfPosition(
+              sourceFile,
+              node.getStart() // Get start position of the call expression
+            );
+
+            // --- Extract Arguments ---
+            const argsText: string[] = [];
+            let errorMessage: string | undefined;
+            let assertionId: string | undefined;
+            if (node.arguments && node.arguments.length > 0) {
+              node.arguments.forEach((arg: ts.Expression) => {
+                // Get the source text of the argument node
+                argsText.push(arg.getText(sourceFile));
+
+                if (ts.isStringLiteral(arg)) {
+                  errorMessage = arg.getText(sourceFile);
+                }
+                else if (ts.isNumericLiteral(arg)) {
+                  assertionId = arg.getText(sourceFile);
+                }
+              });
+            }
+            // --- End Extract Arguments ---
+
+            // Store the information (add 1 to line/char for 1-based indexing)
+            foundCalls.push({
+              fileName: filePath, // Store the full path
+              functionName,
+              line: line + 1,
+              character: character + 1,
+              argumentsText: argsText, // Store the extracted arguments,
+              errorMessage,
+              assertionId: assertionId ?? "INVALID",
+            });
+          }
+        }
+
+        // Continue traversing down the AST
+        ts.forEachChild(node, visit);
+      };
+
+      // Start traversal from the root SourceFile node
+      visit(sourceFile);
+
+    } catch (error: any) {
+      console.error(`Error processing file ${filePath}: ${error.message}`);
```

Comment:
Again, strongly consider just letting the exception bubble up and terminate the app. This "should" never fail, and if it does, IMO it should fail loudly.

Reply:
Good call. Removed the try catch.